waddrmgr/address.go

```diff
@@ -645,9 +667,131 @@ func newScriptAddress(m *ScopedKeyManager, account uint32, scriptHash,
 	}
 
 	return &scriptAddress{
-		manager:         m,
-		account:         account,
-		address:         address,
-		scriptEncrypted: scriptEncrypted,
+		baseScriptAddress: baseScriptAddress{
+			manager:         m,
+			account:         account,
+			scriptEncrypted: scriptEncrypted,
+		},
+		address: address,
+	}, nil
+}
+
+// witnessScriptAddress represents a pay-to-witness-script-hash address.
+type witnessScriptAddress struct {
+	baseScriptAddress
+	address btcutil.Address
+
+	// witnessVersion is the version of the witness script.
+	witnessVersion byte
+
+	// isSecretScript denotes whether the script is considered to be "secret"
+	// and encrypted with the script encryption key or "public" and
+	// therefore only encrypted with the public encryption key.
+	isSecretScript bool
+}
+
+// AddrType returns the address type of the managed address. This can be used
+// to quickly discern the address type without further processing
+//
+// This is part of the ManagedAddress interface implementation.
+func (a *witnessScriptAddress) AddrType() AddressType {
+	return WitnessScript
+}
+
+// Address returns the btcutil.Address which represents the managed address.
+// This will be a pay-to-witness-script-hash address.
+//
+// This is part of the ManagedAddress interface implementation.
+func (a *witnessScriptAddress) Address() btcutil.Address {
+	return a.address
+}
+
+// AddrHash returns the script hash for the address.
+//
+// This is part of the ManagedAddress interface implementation.
+func (a *witnessScriptAddress) AddrHash() []byte {
+	return a.address.ScriptAddress()
+}
+
+// Compressed returns true since witness script addresses are always compressed.
+//
+// This is part of the ManagedAddress interface implementation.
+func (a *witnessScriptAddress) Compressed() bool {
+	return true
+}
+
+// Used returns true if the address has been used in a transaction.
+//
+// This is part of the ManagedAddress interface implementation.
+func (a *witnessScriptAddress) Used(ns walletdb.ReadBucket) bool {
+	return a.manager.fetchUsed(ns, a.AddrHash())
+}
+
+// Script returns the script associated with the address.
+//
+// This is part of the ManagedAddress interface implementation.
+func (a *witnessScriptAddress) Script() ([]byte, error) {
+	// No script is available for a watching-only address manager.
+	if a.isSecretScript && a.manager.rootManager.WatchOnly() {
+		return nil, managerError(ErrWatchingOnly, errWatchingOnly, nil)
+	}
+
+	a.manager.mtx.Lock()
+	defer a.manager.mtx.Unlock()
+
+	// Account manager must be unlocked to decrypt the script.
+	if a.isSecretScript && a.manager.rootManager.IsLocked() {
+		return nil, managerError(ErrLocked, errLocked, nil)
+	}
+
+	cryptoKey := a.manager.rootManager.cryptoKeyScript
```

Comment:
This key isn't new. It existed before for encrypting P2SH scripts. But since it is not available in watch-only wallets, we add a switch to use the "public" encryption key instead of this "secret" script encryption key.